operator/pkg/ciliumendpointslice/manager.go

```diff
@@ -0,0 +1,635 @@
+)
+
+// cesTracker holds the desired state of CiliumEndpointSlice and list of ceps to be removed
+// in next sync with k8s-apiserver.
+type cesTracker struct {
+	// Mutex to protect cep insert/removal in ces and removedCEPs
+	backendMutex lock.RWMutex
+	// The desired state of ces object
+	ces *cilium_v2.CiliumEndpointSlice
+	// set of CEPs to be removed in the CES object in next sync.
+	removedCEPs map[string]struct{}
+	// number of CEPs inserted in a CES
+	cepInserted int64
+	// number of CEPs removed from a CES
+	cepRemoved int64
+	// CES insert time at workqueue
+	cesInsertedAt time.Time
+}
+
+// cesManager
+type cesManager interface {
+	// External APIs to Insert/Remove CEP in local dataStore
+	InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint, ns string) string
+	RemoveCEPFromCache(cepName string, baseDelay time.Duration)
+	// Supporting APIs to Insert/Remove CEP in local dataStore and effectively
+	// manages CES's.
+	getCESFromCache(cesName string) (*cilium_v2.CiliumEndpointSlice, error)
+	getCESCopyFromCache(cesName string) (*cilium_v2.CiliumEndpointSlice, error)
+	updateCESInCache(ces *cilium_v2.CiliumEndpointSlice, deepCopy bool)
+	deleteCESFromCache(cesName string)
+	getRemovedCEPs(string) map[string]struct{}
+	clearRemovedCEPs(string, map[string]struct{})
+	createCES(cesName string) *cesTracker
+	addCEPtoCES(ces *cilium_v2.CoreCiliumEndpoint, cesName string)
+	insertCESInWorkQueue(ces *cesTracker, baseDelay time.Duration)
+	// APIs to collect metrics of CES and CEP
+	getTotalCEPCount() int
+	getCEPCountInCES(cesName string) int
+	getCESCount() int
+	getAllCEPNames() []string
+	getCESMetricCountersAndClear(cesName string) (cepInsert int64, cepRemove int64)
+	getCESQueueDelayInSeconds(cesName string) (diff float64)
+}
+
+// cesMgr is used to batch CEP into a CES, based on FirstComeFirstServe. If a new CEP
+// is inserted, then the CEP is queued in any one of the available CES. CEPs are
+// inserted into CESs without any preference or any priority.
+type cesMgr struct {
+
+	// desiredCESs is used to map CESName to CESTracker[i.e. list of CEPs],
+	// as well as CEPName to CESName.
+	desiredCESs *CESToCEPMapping
+
+	// workqueue is used to sync CESs with the api-server. this will rate-limit the
+	// CES requests going to api-server, ensures a single CES will not be proccessed
+	// multiple times concurrently, and if CES is added multiple times before it
+	// can be processed, this will only be processed only once.
+	queue workqueue.RateLimitingInterface
+
+	// maxCEPsInCES is the maximum number of CiliumCoreEndpoint(s) packed in
+	// a CiliumEndpointSlice Resource.
+	maxCEPsInCES int
+}
+
+// cesManagerFcfs use cesMgr by design, it inherits all the methods from the base cesMgr and there is no
+// special handling required for cesManagerFcfs.
+// cesManagerFcfs indicates ciliumEndpoints are batched based on FirstComeFirtServe algorithm.
+// refer cesMgr comments for more information.
+type cesManagerFcfs struct {
+	cesMgr
+}
```

Comment:
Does `cesManagerFcfs` need to exist if it simply inherits from `cesMgr`? Could we collapse `cesMgr` with `cesManagerFcfs`?

Reply:
The main reason to keep this seperate for `cesManagerFcfs` is, in future if we need something tuned only for `cesManagerFcfs` it's very simple to make changes without touching other implementations.

Reply:
Ah, fcfs is FIFO? I didn't know FCFS but I knew FIFO :)

Reply:
Incoming CEP's are batched into one of the available CES. There is no ordering or priority to choose available CES. The CES are picked based on `FirstComeFirstServe` basis.